### PR TITLE
Star burst Sandcastle example

### PR DESCRIPTION
**Pull request by bagnell**
Adds star burst Sandcastle example.

### CODE REVIEW

**Comment by pjcozzi** [comment]
Update CHANGES.md.  I know this is only a Sandcastle example, but it will be of interest to many users.

**Comment by pjcozzi** [comment]
Are we sure that we want the "Label10" to show up when the billboard is first clicked?

**Comment by pjcozzi** [comment]
Why do the labels flicker when mousing over a billboard?  Are they z-fighting with it?  Should we apply a small eye offset?

**Comment by pjcozzi** [comment]
Given the common horizon view, should we default this to the top hemisphere as we discussed offline?

**Comment by pjcozzi** [comment]
Are we sure that disabling this when the user clicks elsewhere is desirable?  It is the first thing I want to do after expanding it.  It's fine with me if users expect that, but I feel like it would be better with an explicit collapse button.

File: Apps/Sandcastle/gallery/Star Burst.html [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+<script id="cesium_sandcastle_script">
+function startup(Cesium) {
+"use strict";
+//Sandcastle_Begin
+var viewer = new Cesium.Viewer('cesiumContainer', {
+    selectionIndicator : false
+});
+
+// Add labels clustered at the same location
+var numBillboards = 28.0;
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/Star Burst.html]: `28`.  This is an integer, right?

**Comment by pjcozzi** [comment]
Can we evenly spread out the last ring?  Here it is with `numBillboards = 64`:

image

**Comment by pjcozzi** [comment]
Should the billboards have a small eye offset so they are always in front of the polylines?

File: Apps/Sandcastle/gallery/Star Burst.html [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+var radius;
+var centerPosition;
+
+// Additional pixel padding separating the billboards
+var padding = 10.0;
+
+var handler = new Cesium.ScreenSpaceEventHandler(scene.canvas);
+
+// Star burst on left mouse click.
+handler.setInputAction(function(movement) {
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/Star Burst.html]: Can we move the logic out of the input handler like `undoStarBurst`?  Then put the input handlers at the top.  This way it is more clear for users to copy and paste into their specific inputs.

File: Apps/Sandcastle/gallery/Star Burst.html [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+            var angle = 0.0;
+            var angleIncrease;
+            var magnitude;
+            var magIncrease;
+            var maxDimension;
+
+            // Drill pick gets all of the entities under the mouse pointer.
+            // Find the billboards and set their pixel offsets in a circle pattern.
+            for (var i = 0; i < pickedObjects.length; ++i) {
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/Star Burst.html]: `var length = pickedObjects.length`.

Comment (pjcozzi) [inline Apps/Sandcastle/gallery/Star Burst.html]: `;`

ha.

File: Apps/Sandcastle/gallery/Star Burst.html [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+            text : 'Label' + i,
+            show : false
+        }
+    });
+}
+
+var scene = viewer.scene;
+
+// State saved across mouse click and move events
+var pickedEntities;
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/Star Burst.html]: Put this into an object literal.

In general, think about making this as easy as possible for users to copy and paste into their project.  Perhaps the whole thing could even go into a class that one day could be a candidate for core Cesium.

**Comment by pjcozzi** [comment]
Looks good!

Have you tried it with billboards of vastly different sizes?  It is totally out of scope for this PR, but I imagine there would be sorting heuristics, for example, like texture atlas packing, to make it look good.

**Comment by bagnell** [comment]
>Are we sure that disabling this when the user clicks elsewhere is desirable?

Yes, that's how GE works too. Plus that would ruin the illusion of how I did the lines. To see what I mean, just comment out the `undoStarBurst` call in the camera move event.

**Comment by bagnell** [comment]
>Given the common horizon view, should we default this to the top hemisphere as we discussed offline?

Sure. It doesn't look as bad as I thought. Let me know what you think.

**Comment by bagnell** [comment]
>Should the billboards have a small eye offset so they are always in front of the polylines?

Unfortunately I can't do that for the same reason I collapse the burst on camera move. The lines are drawn from the center to the near plane where the billboard is drawn. The reason is I didn't want to keep track of what frustum the billboard is rendered in the get the depth and reconstruct its position. I can't just call `pickPosition` because billboards are rendered in the transparent pass and don't write depth regardless of what the given render state is.

**Comment by pjcozzi** [comment]
Looks OK.  Why can't it fill 270 to 90 degrees.  Looks like it doesn't include 270.

image

**Comment by bagnell** [comment]
>Looks OK. Why can't it fill 270 to 90 degrees. Looks like it doesn't include 270.

Expect it to fit exactly? I could draw them there but then some billboards would draw partly past 270.

**Comment by bagnell** [comment]
>Are we sure that we want the "Label10" to show up when the billboard is first clicked?

That seems to always happen on the first pick, but after that its hard to reproduce consistently. I'm not sure why.

**Comment by bagnell** [comment]
@pjcozzi This is ready for another review.

Two things haven't been done. I set the eye offset for the billboards to appear in front of the lines but it doesn't seem to have any effect. I'm also unsure what is causing the label to appear sometimes on the first click.

**Comment by pjcozzi** [comment]
>     // Remove the star burst if the mouse exits the screen space circle.
    // If the mouse is inside the circle, show the label of the billboard the mouse is hovering over.

This feels too aggressive.  If I move the mouse below the star burst it goes away.  Perhaps the right approach is to use a full circle and slightly scale the radius in case the user overshoots.

**Comment by pjcozzi** [comment]
Is this supposed to take into account scale?

```
scale : 6 * (Math.random())
```

image


Should it also reset/restore pixel offset and eye offset x/y?

**Comment by pjcozzi** [comment]
For the eye offset, the billboard eye offset "works" in that, for example:
```
    entity.billboard.eyeOffset =
       new Cesium.Cartesian3(0.0, 0.0, -100000000.0);
```
brings the billboards in front of the near plan so they disappear (likewise `z = 100000000.0` puts them behind the globe).  Perhaps look more closely at the polyline vertices.  Looks like it should be testing depth.

**Comment by pjcozzi** [comment]
Also merge in master when you have a chance.

**Comment by bagnell** [comment]
>Is this supposed to take into account scale?

It takes into account scale, but not different scales or different size billboards. Since we are using `drillPick` to get all of the billboards, its possible to click part of a large billboard, have some of the in the burst and still have some overlapping at the center.

**Comment by bagnell** [comment]
@pjcozzi This is ready for another look.

**Comment by pjcozzi** [comment]
Looks great!

CC @GatorScott

**Pull request merged**